### PR TITLE
Fix #4 & enhanced RedShift functionality

### LINKED ISSUE #4: Existing RedShift subnet group not honored
I may be using the module in the incorrect way, but I believe the module currently does not honor an existing RedShift subnet group.

## Expected Behavior
Based on this [module's documentation](https://registry.terraform.io/modules/terraform-aws-modules/redshift/aws/1.0.0?tab=inputs), I think I should do one of the following:

- Specify the name of an existing RedShift subnet group in the input variable `redshift_subnet_group_name` from the VPC module using `${module.vpc.redshift_subnet_group}` (or created myself in some way), **_OR..._**
- Specify a list of subnets using the `subnets` input, whereby a RedShift subnet group is then created for me by this module.

In other words, I tell the module to use an existing subnet group (which already has defined subnets inside of it), or I specify subnet IDs in `subnets`, and the module creates a subnet group for me.

## Current Behavior

The module tries to create a new subnet group even if you specify an existing subnet group.

Similar to the example at `examples/complete/main.tf`, I am using the official AWS Terraform VPC module (https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws). When you specify the following block in that module...

```hcl
redshift_subnets = ["10.10.41.0/24", "10.10.42.0/24", "10.10.43.0/24"]
```

...it will create the subnets you specify, as well as a RedShift subnet group for you; the output from the VPC module is called `redshift_subnet_group`, which is also used in the example in this repository.

Here is my example module config:

```hcl
module "redshift" {
    cluster_database_name = "mydatabase"
    cluster_identifier = "mycluster"
    cluster_master_password = "Securepassword1"
    cluster_master_username = "redshiftadmin"
    cluster_node_type = "dc2.large"
    cluster_number_of_nodes = 1
    redshift_subnet_group_name = "${module.vpc.redshift_subnet_group}"
    source  = "terraform-aws-modules/redshift/aws"
    version = "1.0.0"
    vpc_security_group_ids = [
        "${module.vpc.default_security_group_id}"
    ]
}
```
Even though I specify which RedShift subnet group I want to use, running `terraform plan` results in the following:

```hcl
  + module.redshift.aws_redshift_cluster.this
      id:                                  <computed>
      allow_version_upgrade:               "true"
      automated_snapshot_retention_period: "0"
      availability_zone:                   <computed>
      bucket_name:                         <computed>
      cluster_identifier:                  "mycluster"
      cluster_parameter_group_name:        <computed>
      cluster_public_key:                  <computed>
      cluster_revision_number:             <computed>
      cluster_security_groups.#:           <computed>
      cluster_subnet_group_name:           "my-existing-redshift-subnet-group"
      cluster_type:                        <computed>
      cluster_version:                     "1.0"
      database_name:                       "mydatabase"
      enable_logging:                      <computed>
      encrypted:                           "false"
      endpoint:                            <computed>
      enhanced_vpc_routing:                <computed>
      kms_key_id:                          <computed>
      master_password:                     <sensitive>
      master_username:                     "redshiftadmin"
      node_type:                           "dc2.large"
      number_of_nodes:                     "1"
      port:                                "5439"
      preferred_maintenance_window:        "sat:10:00-sat:10:30"
      publicly_accessible:                 "false"
      s3_key_prefix:                       <computed>
      skip_final_snapshot:                 "true"
      vpc_security_group_ids.#:            "1"
      vpc_security_group_ids.4287799107:   "sg-07773f214e065ee54"

  + module.redshift.aws_redshift_subnet_group.this
      id:                                  <computed>
      description:                         "Redshift subnet group of mycluster"
      name:                                "mycluster"
```
As you can see, the module tries to create `module.redshift.aws_redshift_subnet_group.this`. Even weirder, it appears as if the module **_did_** correctly identify my existing subnet group in the following line:

```
      cluster_subnet_group_name:           "my-existing-redshift-subnet-group"
```
However, when I finally run a `terraform apply`, I get the following error:

```
Error: Error applying plan:

1 error(s) occurred:

* module.redshift.aws_redshift_subnet_group.this: 1 error(s) occurred:

* aws_redshift_subnet_group.this: Error creating Redshift Subnet Group: InvalidParameterValue: Missing subnetIds.
	status code: 400, request id: 0c382a32-4987-11e8-8c2c-b7112d904d32
```

This is because I'm not specifying `subnets` since I want to use an already existing subnet group.

## Possible Solution

It appears that this line (https://github.com/terraform-aws-modules/terraform-aws-redshift/blob/master/main.tf#L21) gets things right; the reason is because, like my output above, it correctly identifies my already existing subnet group name when the `plan` is executed.

Somehow in the following code block (https://github.com/terraform-aws-modules/terraform-aws-redshift/blob/master/main.tf#L57-L65) the resource...

```
count = "${local.enable_create_redshift_subnet_group}"
```

Is accidentally getting evaluated as `1` when it should be `0`.

**_EDIT:_** Fixed with this commit => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/be8eeabd743af289e1eefd8d002272fb672466c5

## Steps to Reproduce

1. Manually create a RedShift subnet group in AWS.
2. Run the code snippet I've placed above, replacing the `redshift_subnet_group_name` with the name you just used in step 1. Make sure you're using the versions specified below for Terraform, Terraform AWS provider, and this module.
3. Run `terraform apply`
4. Observe that it attempts to create another subnet group even though you've already specified one.

## Context (Environment)
Terraform v0.11.7
Terraform-aws-modules/redshift/aws V1.0.0
Provider AWS v1.16

### PR DESCRIPTION
**Pull request by sc250024**
Fixes issue https://github.com/terraform-aws-modules/terraform-aws-redshift/issues/4, and adds other functionality to the RedShift module.

## Description
- Fix for https://github.com/terraform-aws-modules/terraform-aws-redshift/issues/4 => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/e1ed24f3ef860fbae861daca2cae865058d40293. The logic is this: A parameter group should not be created if `parameter_group_name` is not specified since RedShift will create a default parameter group for you. However, if `redshift_subnet_group_name` is not specified, then the resource `redshift.aws_redshift_subnet_group.this` should be created.
- Enabling enhanced VPC routing => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/019ff28a4f0c6466f5d79eefbfa7e296202f9711
- Calculating the value of `cluster_type` based on the number of nodes the user specifies => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/0b0dc7d4ba4268fd6a3cd1ea8abf71f8b8ddeb6f and https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/fca695d78c673a44ffc7af48ede3f4d1eb6bce40
- Added variable `allow_version_upgrade` => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/fc6c31e01d161b57661805a2aedaeb4cad321a3d
- Added variable `final_snapshot_identifier` => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/7c86caaaaa1bd4255aad15ba6de2202381048c1f
- Added S3 logging => https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/5792606cbd6fded11502908c0d9b1a638c40b156

## Motivation and Context
Initially I was attempting to solve https://github.com/terraform-aws-modules/terraform-aws-redshift/issues/4, and noticed there was some other functionality I would like included in the module as well.

## How Has This Been Tested?
Tested locally with AWS VPC creation of RedShift resources.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -55,7 +65,7 @@ resource "aws_redshift_parameter_group" "this" {
 }
 
 resource "aws_redshift_subnet_group" "this" {
-  count = "${local.enable_create_redshift_subnet_group}"
+  count = "${var.redshift_subnet_group_name != "" ? 0 : 1}"
```
Comment (antonbabenko) [inline main.tf]: It should be `==` instead of `!=`

File: main.tf [hunk 2]
```diff
@@ -37,13 +38,22 @@ resource "aws_redshift_cluster" "this" {
 
   tags = "${var.tags}"
 
+  # Enhanced VPC routing
+  enhanced_vpc_routing = "${var.enhanced_vpc_routing}"
+
   lifecycle {
```
Comment (antonbabenko) [inline main.tf]: Move `tags` and `lifecycle` arguments at the end as they were before.

File: main.tf (hunk 1 above)
Comment (sc250024) [reply main.tf]: Come to think of it, should we change the `locals` to be similar to the Terraform RDS module? https://github.com/terraform-aws-modules/terraform-aws-rds/blob/master/main.tf#L1-L6

Comment (antonbabenko) [reply main.tf]: Yes, using `locals` is much better.

File: main.tf (hunk 2 above)
Comment (sc250024) [reply main.tf]: Done

File: main.tf (hunk 1 above)
Comment (sc250024) [reply main.tf]: The way it was implemented before, using the blocks:
```
  enable_create_redshift_subnet_group    = "${var.redshift_subnet_group_name == "" ? 0 : 1}"
  enable_create_redshift_parameter_group = "${var.parameter_group_name == "" ? 0 : 1}"
```

Was actually was caused the problem. If you look at my logs in https://github.com/terraform-aws-modules/terraform-aws-redshift/issues/4, for some reason, using `enable_create_redshift_` the way it was written resulted in subnet groups being created even when a user specified an already existing subnet group.

In the RDS module, the difference is that the code block:

```
  enable_create_db_subnet_group    = "${var.db_subnet_group_name == "" ? var.create_db_subnet_group : 0}"
  enable_create_db_parameter_group = "${var.parameter_group_name == "" ? var.create_db_parameter_group : 0}"
```

Has explicit `create_db_*` variables. Do we want equivalent `create_redshift_parameter_group` and `create_redshift_subnet_group` variables in this module?

**Comment by sc250024** [comment]
@antonbabenko My latest commit https://github.com/terraform-aws-modules/terraform-aws-redshift/pull/5/commits/be8eeabd743af289e1eefd8d002272fb672466c5 restores most of the restores **_most_** of the functionality for the locals `enable_create_redshift_subnet_group` and `enable_create_redshift_parameter_group` with one important change. The logic is this:

- If `parameter_group_name` is blank, the default behavior **_should not_** be to create a custom parameter group since AWS will create a default one for you. The resource `aws_redshift_parameter_group.this` should only be created when the user wants to specify a custom parameter group.

- Conversely, if `redshift_subnet_group_name` is blank, the resource `aws_redshift_subnet_group` **_should_** be created (and thus, the user should also specify `subnets`) since a subnet group needs to exist in order to spin up the RedShift cluster.

Hopefully this resolves things, and can be merged now.

**Comment by antonbabenko** [comment]
Took me some time to think about it and understand it. Good job! Merging now.

**Comment by antonbabenko** [comment]
[v1.1.0](https://github.com/terraform-aws-modules/terraform-aws-redshift/releases/v1.1.0) has been released.

**Pull request merged**